From 834f1c4037ad430f38c1baf7addf885ea1471ebd Mon Sep 17 00:00:00 2001
From: Miranda Kephart <mkephart@google.com>
Date: Sun, 27 Aug 2023 20:33:40 +0530
Subject: [PATCH] Fix for Update quickshare intent rather than  recreating

Currently, we extract the quickshare intent and re-wrap it as a new
PendingIntent once we get the screenshot URI. This is insecure as
it leads to executing the original with SysUI's permissions, which
the app may not have. This change switches to using Intent.fillin
to add the URI, keeping the original PendingIntent and original
permission set.

Bug: 278720336
Test: manual (to test successful quickshare), atest
SaveImageInBackgroundTaskTest (to verify original pending intent
unchanged)
---
 .../systemui/screenshot/SaveImageInBackgroundTask.java        | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/screenshot/SaveImageInBackgroundTask.java b/packages/SystemUI/src/com/android/systemui/screenshot/SaveImageInBackgroundTask.java
index e02912fa18ed..3ea09b7eddcf 100644
--- a/packages/SystemUI/src/com/android/systemui/screenshot/SaveImageInBackgroundTask.java
+++ b/packages/SystemUI/src/com/android/systemui/screenshot/SaveImageInBackgroundTask.java
@@ -186,7 +186,7 @@ class SaveImageInBackgroundTask extends AsyncTask<Void, Void, Void> {
                     smartActionsEnabled);
             mImageData.quickShareAction = createQuickShareAction(
                     mQuickShareData.quickShareAction, mScreenshotId, uri, mImageTime, image,
-		    user);
+		    user, smartActionsEnabled);
             mImageData.subject = getSubjectString();
 
             mParams.mActionsReadyListener.onActionsReady(mImageData);
@@ -433,7 +433,7 @@ class SaveImageInBackgroundTask extends AsyncTask<Void, Void, Void> {
     @VisibleForTesting
     Notification.Action createQuickShareAction(
 	    Notification.Action quickShare, String screenshotId, Uri uri, long imageTime,
-            Bitmap image, UserHandle user) {
+            Bitmap image, UserHandle user, boolean mSmartActionsEnabled) {
         if (quickShare == null) {	    
             return null;
 	} else if (quickShare.actionIntent.isImmutable()) {
-- 
2.17.1

